### DISCUSSION TITLE
Define an argument with default options and a completing-read-multiple reader, which retains selections between readings (using :always-read)?

### DISCUSSION THREAD

**Discussion by Hugo-Heagren** [magit/transient#194] [Support] [answered]
Sorry for the long title.

I don't understand how to do what I want, and transient is young and specific enough that I thought this (rather than reddit, or emacs-devel) would be the best place to ask. Hope that's right.

I want to define a transient argument which:
- reads with completing-read-multiple (hereafter 'crm')
- generates a completion table for crm dynamically (e.g. by checking a user option, or by calling a function stored in that option)
- has `:always-read` set to `t` (it's still possible to unset it, just read, and in the crm session return nothing)
- **retains crm candidate choices between read sessions**, such that if I edit the option, set it "foo,bar", then go to set it again, the initial input to crm that time round will be "foo,bar," (note the trailing crm-separator, so that crm recognises this is a list of whole elements without a trailing completion string).

I've got as far as something like this:

```elisp
(transient-define-argument magit-send-email:-b ()
  :description "Bcc"
  :class 'transient-option
  :key "-b"
  :argument "--bcc="
  :reader #'completing-read-multiple
  :multi-value t
  :always-read t
  :prompt "Bcc: ")
```

`transient-option` seems like the right choice, because the value of the option is a list (of arbitrary length) of strings, which the user specifies. 

As far as I can tell there are two problems.

First, transient's readers take `(prompt init-value history)` as arguments, so I don't know how to specify the choices. I tried setting `:choices` but that didn't seem to do anything. Obviously I could write a lamba wrapper over crm which includes the choices, but I'm guessing there's a more elegant way?

First question: how I can I dynamically (in the above sense) set the choices for the crm reader?

Secondly, I need to set the initial value for crm. I tried setting `:init-value` to a function which gets the current state of argument, but I couldn't get it to work at all, even when for debugging I set it a function which just always returned a string (or list of strings). I tried setting a lambda wrapper over crm (with the arguments in the above form) to make sure the args went in the right places and it still didn't work. I think part of the problem is  there is no `transient-init-value` implementation for `transient-option`, and [the one it presumably inherits from `transient-argument`](https://github.com/magit/transient/blob/master/lisp/transient.el#L2686) doesn't check the `init-value` slot in the object. IIUC, that means the value of `:init-value` will just be ignored, which might be my problem?

Second question: how can I dynamically (in a similar sense) set the initial value for every crm session?

**Comment by tarsius** [comment, marked as answer]
> Obviously I could write a lamba wrapper over crm which includes the choices, but I'm guessing there's a more elegant way?

Well, maybe make it a named function but other than that this is what you are expected to do. And why not? The place where the completion happens seems like the prefect place to specify the details of how the completion should be performed. Putting the information in the slots of an object/class just spreads the relevant information further.

Of course `transient` already does the latter but that is done to reduce the amount of boilerplate. So the things that are very similar between all or most infix arguments are controlled by setting the values of certain slots, but the things that are very different between different suffixes should be placed in just one location: the respective reader function.

An additional benefit of that is that you can use the specialized reader for multiple related suffixes. All you have to do for each suffix is to set only the reader slot, instead of having to set multiple slots to potentially non-trivial values.

Here's how how magit reads the files that go after `--`:

```elisp
(transient-define-argument magit:-- ()
  :description "Limit to files"
  :class 'transient-files
  :key "--"
  :argument "--"
  :prompt "Limit to file(s): "
  :reader #'magit-read-files
  :multi-value t)

(defun magit-read-files (prompt initial-input history &optional list-fn)
  (magit-completing-read-multiple* prompt
                                   (funcall (or list-fn #'magit-list-files))
                                   nil nil
                                   (or initial-input (magit-file-at-point))
                                   history))
```

Grep magit for `:reader` to find many more examples.

If you are wondering why `magit-read-files` takes a forth argument and how that can possibly works, then search for uses of that function too.
